app/services/foreman_bootdisk/iso_generator.rb

```diff
@@ -15,6 +15,10 @@ class ISOGenerator
     def self.generate_full_host(host, opts = {}, &block)
       raise Foreman::Exception.new(N_('Host is not in build mode, so the template cannot be rendered')) unless host.build?
 
+      # update build token
+      token = host.token
+      token.update(expires: token.expires + Setting[:token_duration].minutes)
+
```

Comment:
This code will fail when you try to build a bootdisk on a host that has no token (e.g. host which is not in build mode). In that case `host.token` is `nil` which will lead to NilException.

Also, you cannot create a token just by creating it, it would create new token without orchestration.

I think what we want to do here is: 

* if the host is in build mode and it has expired token
* then create a new one

Please use host's methods set_token, token_expired? and build? - see for example in here: https://github.com/theforeman/foreman/blob/develop/app/models/concerns/hostext/token.rb

Reply:
Few questions, why do we need orchestration for token creation? I think the necessary orchestration is rather driven by `set_build` on host. Note that the host needs to be in the build mode for user to be able to download the bootdisk. When they try to get the bootdisk for hosts which is not in build mode, they receive (now ugly) error message about the fact.

IMHO the correct behavior is to prolong the existing token whenever a bootdisk is generated or create a new token, while keeping the check for build mode being enabled (meaning all provisioning services are prepare for provisioning)

Reply:
What I understand this bug, the problem is with hosts that are prepared for building because they are missed the chance to provision due to token expiration. I know that prolonging token expiration is really bad security issue like in resetting password etc. However we can't do that because token is written into image so prolonging token is one logic solution for that. 

And as @ares I think that orchestration is not necessary. At least, this issue address to the hosts that has prepared orchestration (because build mode).

Reply:
Right, prolonging is also an option. In that case, just update the code so it won't break inc ase token is `nil` and we are good to go.